uniffi_bindgen/src/interface/enum_.rs

```diff
@@ -31,34 +31,84 @@
 //! let e = ci.get_enum_definition("Example").unwrap();
 //! assert_eq!(e.name(), "Example");
 //! assert_eq!(e.variants().len(), 2);
-//! assert_eq!(e.variants()[0], "one");
-//! assert_eq!(e.variants()[1], "two");
+//! assert_eq!(e.variants()[0].name(), "one");
+//! assert_eq!(e.variants()[1].name(), "two");
+//! # Ok::<(), anyhow::Error>(())
+//! ```
+//!
+//! Like in Rust, UniFFI enums can contain associated data, but this needs to be
+//! declared with a different syntax in order to work within the restrictions of
+//! WebIDL. A declaration like this:
+//!
+//! ```
+//! # let ci = uniffi_bindgen::interface::ComponentInterface::from_webidl(r##"
+//! # namespace example {};
+//! [Enum]
+//! interface Example {
+//!   Zero();
+//!   One(u32 first);
+//!   Two(u32 first, string second);
+//! };
+//! # "##)?;
+//! # Ok::<(), anyhow::Error>(())
+//! ```
+//!
+//! Will result in an [`Enum`] member whose variants have associated fields:
+//!
+//! ```
+//! # let ci = uniffi_bindgen::interface::ComponentInterface::from_webidl(r##"
+//! # namespace example {};
+//! # [Enum]
+//! # interface ExampleWithData {
+//! #   Zero();
+//! #   One(u32 first);
+//! #   Two(u32 first, string second);
+//! # };
+//! # "##)?;
+//! let e = ci.get_enum_definition("ExampleWithData").unwrap();
+//! assert_eq!(e.name(), "ExampleWithData");
+//! assert_eq!(e.variants().len(), 3);
+//! assert_eq!(e.variants()[0].name(), "Zero");
+//! assert_eq!(e.variants()[0].fields().len(), 0);
+//! assert_eq!(e.variants()[1].name(), "One");
+//! assert_eq!(e.variants()[1].fields().len(), 1);
+//! assert_eq!(e.variants()[1].fields()[0].name(), "first");
 //! # Ok::<(), anyhow::Error>(())
 //! ```
 
-use anyhow::Result;
+use anyhow::{bail, Result};
 
+use super::record::Field;
+use super::types::Type;
 use super::{APIConverter, ComponentInterface};
 
-/// Represents a simple C-style enum, with named variants.
+/// Represents an enum with named variants, each of which may have named
+/// and typed fields.
 ///
-/// In the FFI these are turned into a plain u32, with variants numbered
-/// in the order they appear in the declaration, starting from 1.
+/// Enums are passed across the FFI by serializing to a bytebuffer, with a
+/// i32 indicating the variant followed by the serialization of each field.
 #[derive(Debug, Clone, Hash)]
 pub struct Enum {
     pub(super) name: String,
-    pub(super) variants: Vec<String>,
+    pub(super) variants: Vec<Variant>,
 }
 
 impl Enum {
     pub fn name(&self) -> &str {
         &self.name
     }
-    pub fn variants(&self) -> Vec<&str> {
-        self.variants.iter().map(|v| v.as_str()).collect()
+    pub fn variants(&self) -> Vec<&Variant> {
+        self.variants.iter().collect()
+    }
+
+    pub fn has_associated_data(&self) -> bool {
+        self.variants.iter().any(Variant::has_fields)
```

Comment:
Yeah, I think you're right. I'm not super happy about it because it feels like an implementation detail of the (what we hope to be temporary) IDL leaking all the way through to the generated bindings, but that's probably less bad than having a confusing user experience.

Reply:
I'm updating this to mark enums that come from `enum foo` as "flat" and enums that come from `[Enum] interface` as "not flat", and will use that in codegen to switch between the two implementations.